internal/buffer/buffer.go

```diff
@@ -930,6 +946,101 @@ func (b *Buffer) Write(bytes []byte) (n int, err error) {
 	return len(bytes), nil
 }
 
+func (b *Buffer) updateDiffSync() {
+	b.diffLock.Lock()
+	defer b.diffLock.Unlock()
+
+	b.diff = make(map[int]DiffStatus)
+
+	if b.diffBase == nil {
+		return
+	}
+
+	differ := dmp.New()
+	baseRunes, bufferRunes, _ := differ.DiffLinesToRunes(string(b.diffBase), string(b.Bytes()))
+	diffs := differ.DiffMainRunes(baseRunes, bufferRunes, false)
+	lineN := 0
+
+	for _, diff := range diffs {
+		lineCount := len([]rune(diff.Text))
+
+		switch diff.Type {
+		case dmp.DiffEqual:
+			lineN += lineCount
+		case dmp.DiffInsert:
+			var status DiffStatus
+			if b.diff[lineN] == DSDeletedAbove {
+				status = DSModified
+			} else {
+				status = DSAdded
+			}
+			for i := 0; i < lineCount; i++ {
+				b.diff[lineN] = status
+				lineN++
+			}
+		case dmp.DiffDelete:
+			b.diff[lineN] = DSDeletedAbove
+		}
+	}
+}
+
+// UpdateDiff computes the diff between the diff base and the buffer content.
+// The update may be performed synchronously or asynchronously.
+// UpdateDiff calls the supplied callback when the update is complete.
+// The argument passed to the callback is set to true if and only if
+// the update was performed synchronously.
+// If an asynchronous update is already pending when UpdateDiff is called,
+// UpdateDiff does not schedule another update, in which case the callback
+// is not called.
+func (b *Buffer) UpdateDiff(callback func(bool)) {
+	if b.updateDiffTimer != nil {
+		return
+	}
+
+	lineCount := b.LinesNum()
+	if b.diffBaseLineCount > lineCount {
+		lineCount = b.diffBaseLineCount
+	}
+
+	if lineCount < 1000 {
+		b.updateDiffSync()
+		callback(true)
+	} else if lineCount < 30000 {
+		b.updateDiffTimer = time.AfterFunc(500*time.Millisecond, func() {
```

Comment:
What is the point of the `updateDiffTimer`? This will call the given function in a separate goroutine after 500ms. Wouldn't it be better to simply call the function in a separate goroutine right away? Something like

```go
go func() {
    b.updateDiffSync()
    callback(false)
}()
```

and then get rid of `updateDiffTimer` entirely?

Reply:
This is a debounce, a rate-limiting mechanism. It consolidates multiple calls to `UpdateDiff` made within a short period of time into a single diff operation. Note that `UpdateDiff` returns immediately if a timer is active (i.e., if a diff operation is already scheduled).

As an example, imagine you are editing a file with 20,000 lines (for which a diff operation takes about 30 milliseconds) and typing 10 characters per second. Without the debounce, Micro would perform 10 diff operations per second, taking 300 milliseconds of core time, or 30% of a single CPU core. Additionally, each of these operations would schedule a redraw once it is complete. With the debounce, Micro performs only 2 diff operations and redraws per second, using just 6% of one CPU core.

In other words, this code substantially reduces the CPU usage, at the small price of having to wait a maximum of 0.5 seconds to see diff updates.